File: tests/algorithms/x/Pascal/maths/combinations.pas
{$mode objfpc}{$modeswitch nestedprocvars}
program Main;
uses SysUtils;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
procedure error(msg: string);
begin
  panic(msg);
end;
function _to_float(x: integer): real;
begin
  _to_float := x;
end;
function to_float(x: integer): real;
begin
  to_float := _to_float(x);
end;
procedure json(xs: array of real);
var i: integer;
begin
  write('[');
  for i := 0 to High(xs) do begin
    write(xs[i]);
    if i < High(xs) then write(', ');
  end;
  writeln(']');
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  k: integer;
  n: integer;
function combinations(n: integer; k: integer): integer; forward;
function combinations(n: integer; k: integer): integer;
var
  combinations_res: integer;
  combinations_i: integer;
begin
  if (k < 0) or (n < k) then begin
  panic('Please enter positive integers for n and k where n >= k');
end;
  combinations_res := 1;
  combinations_i := 0;
  while combinations_i < k do begin
  combinations_res := combinations_res * (n - combinations_i);
  combinations_res := combinations_res div (combinations_i + 1);
  combinations_i := combinations_i + 1;
end;
  exit(combinations_res);
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  writeln('The number of five-card hands possible from a standard fifty-two card deck is: ' + IntToStr(combinations(52, 5)));
  writeln('');
  writeln(('If a class of 40 students must be arranged into groups of 4 for group projects, there are ' + IntToStr(combinations(40, 4))) + ' ways to arrange them.');
  writeln('');
  writeln(('If 10 teams are competing in a Formula One race, there are ' + IntToStr(combinations(10, 3))) + ' ways that first, second and third place can be awarded.');
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.

File: tests/algorithms/transpiler/FS/maths/combinations.fs
// Generated 2025-08-17 08:49 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let rec _str v =
    match box v with
    | :? float as f -> sprintf "%.15g" f
    | _ ->
        let s = sprintf "%A" v
        s.Replace("[|", "[")
         .Replace("|]", "]")
         .Replace("; ", " ")
         .Replace(";", "")
         .Replace("\"", "")
let _floordiv (a:int) (b:int) : int =
    let q = a / b
    let r = a % b
    if r <> 0 && ((a < 0) <> (b < 0)) then q - 1 else q
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec combinations (n: int) (k: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable n = n
    let mutable k = k
    try
        if (k < 0) || (n < k) then
            ignore (failwith ("Please enter positive integers for n and k where n >= k"))
        let mutable res: int = 1
        let mutable i: int = 0
        while i < k do
            res <- res * (n - i)
            res <- _floordiv (int res) (int (i + 1))
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
ignore (printfn "%s" ("The number of five-card hands possible from a standard fifty-two card deck is: " + (_str (combinations (52) (5)))))
ignore (printfn "%s" (""))
ignore (printfn "%s" (("If a class of 40 students must be arranged into groups of 4 for group projects, there are " + (_str (combinations (40) (4)))) + " ways to arrange them."))
ignore (printfn "%s" (""))
ignore (printfn "%s" (("If 10 teams are competing in a Formula One race, there are " + (_str (combinations (10) (3)))) + " ways that first, second and third place can be awarded."))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
